### PR TITLE
raftstore-v2: implement a simplified version of CoprocessorHost

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
Signed-off-by: SpadeA-Tang <u6748471@anu.edu.au>

### What is changed and how it works?

Issue Number: Ref #12842

What's Changed:

```
A simplified version of CoprocessorHost in v1, it conveys information to LockManager.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/raft/peer.rs [hunk 1]
```diff
@@ -641,6 +660,34 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             .advance_apply(apply_index, term, region);
     }
 
+    /// Register self to apply_scheduler so that the peer is then usable.
+    /// Also trigger `RegionChangeEvent::Create` here.
+    pub fn activate<T>(&mut self, ctx: &mut StoreContext<EK, ER, T>) {
```
Comment (tabokie) [inline components/raftstore-v2/src/raft/peer.rs:665]: Why create a function for this, I see it's only called once.

File: components/raftstore-v2/src/raft/peer.rs [hunk 2]
```diff
@@ -641,6 +660,34 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+    }
+
+    #[inline]
+    pub fn set_lead_transferee(&mut self, lead_transferee: u64) {
+        self.lead_transferee = lead_transferee;
+    }
+
+    /// Update states of the peer which can be changed in the previous raft
+    /// tick.
+    pub fn post_raft_group_tick(&mut self) {
```
Comment (tabokie) [inline components/raftstore-v2/src/raft/peer.rs:687]: These two can be merged into `refresh_lead_transferee()`

File: src/server/lock_manager/mod.rs [hunk 3]
```diff
@@ -205,6 +205,11 @@ impl LockManager {
         role_change_notifier.register(host);
     }
 
+    /// Creates a `RoleChangeNotifier` of the deadlock detector worker
+    pub fn generate_notifier(&self) -> RoleChangeNotifier {
```
Comment (tabokie) [inline src/server/lock_manager/mod.rs:209]:
```suggestion
    pub fn new_notifier(&self) -> RoleChangeNotifier {
```

File: src/server/raftkv2/node.rs [hunk 4]
```diff
@@ -85,6 +85,7 @@ where
         snap_mgr: TabletSnapManager,
         concurrency_manager: ConcurrencyManager,
         causal_ts_provider: Option<Arc<CausalTsProviderImpl>>, // used for rawkv apiv2
+        lock_manager_observer: Arc<dyn LockManagerNotifier>,
```
Comment (tabokie) [inline src/server/raftkv2/node.rs:88]: Use Option.

File: components/raftstore-v2/src/raft/peer.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/raft/peer.rs:665]: After the change of split init, yes, it is.

File: src/server/raftkv2/node.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply src/server/raftkv2/node.rs:88]: Why, we can use a dummy notifier if it is not needed (in test for example).

File: components/raftstore-v2/src/raft/peer.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/raft/peer.rs]: Done

File: src/server/raftkv2/node.rs (hunk 4 above)
Comment (tabokie) [reply src/server/raftkv2/node.rs:88]: Clearly None is better than dummy?

**Review by tabokie — APPROVED** [review]

File: src/server/raftkv2/node.rs (hunk 4 above)
Comment (tabokie) [reply src/server/raftkv2/node.rs:88]: Anyway, if previously coprocessor host is not an option, I'm okay with dummy too.

**Pull request merged**